rfcs/0089-spill-hash-aggregation.md

```diff
@@ -0,0 +1,98 @@
+---
+
+feature: Spill Hash Aggregation
+authors:
+  - "DylanChen"
+start_date: "2024/05/13"
+
+---
+
+# Spill Hash Aggregation
+
+## Motivation
+
+RisingWave batch queries are executed purely in memory currently which could easily reach to the the reserved batch memory limit. Hash aggregation is one of the operator that could require a large amount of memory (depending on the cardinality of the group by keys). For example, deduplication is a common feature and in batch queries, hash aggregation is often used to implement it. During deduplication, the in-memory hash table of the hash aggregation could be quite large and cause an OOM error. If we could spill the in-memory hash table to the disk, it would greatly improve the cluster stability and be able to run a large query with a small machine.
+
+## Design
+
+As we know a batch query would be fragmented into different fragments and a fragment would be further assigned a parallelism to execute, for each parallelism, we call a task. Hash aggregation is an executor within a task, aka, HashAggExec. To spill a HashAgg, we need to spill the hash table and the input chunks. Ideally, we want HashAggExec to work normally when memory is sufficient and when memory is not enough, we spill the hash table and all input chunks.
+
+![spill-hash-agg](images/0089-spill-hash-aggregation/spill-hash-agg.png)
+
+### Partitions
+
+First, we need to choose a partition number to partition the hash table and input chunks. After partitioning, theoretically, each partition would only contain 1/partition_num of the original data. If this size could be fitted in the memory, we can process the HashAgg partition by partition. If this size is still too large to be fitted in the memory, we need to recursively apply the spill algorithm. When recursively applying the spill algorithm, we need to make sure they use different hash functions to avoid data skew.
```

Comment:
>  When recursively applying the spill algorithm, we need to make sure they use different hash functions to avoid data skew.

Could you elaborate on this?

Reply:
Why not just use vnode, instead of a separate partition strategy?

Reply:
> Why would we need to recursively apply the spill algorithm?

Because we don't know how much data the input has. Even if we have 20 partitions by default, a single partition data size still could be too large to fit in memory, so we need to spill the partition again, i.e. recursively apply.

Reply:
> Why not just use vnode, instead of a separate partition strategy?
For batch query, we don't always have a vnode. Vnode is only associated with the data within a table, but batch hash join input could be intermediate data.